src/merklized_state/data_source.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the HotShot Query Service library.
+//
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not,
+// see <https://www.gnu.org/licenses/>.
+
+//! Data for the [`merklized_state`](super) API.
+//!
+//! This module facilitates storing the state of a Merkle Tree at a specific point in time
+//! and provides methods for querying and reconstructing the snapshot.
+//!
+use ark_serialize::CanonicalDeserialize;
+use derive_more::{Display, From};
+
+use async_trait::async_trait;
+use commit::Commitment;
+use derivative::Derivative;
+use hotshot::traits::ValidatedState;
+use hotshot_types::{data::Leaf, traits::node_implementation::NodeType};
+use jf_primitives::{
+    circuit::merkle_tree::MembershipProof,
+    merkle_tree::{
+        prelude::{MerkleNode, MerklePath},
+        Element, Index, MerkleTreeScheme, NodeValue, ToTraversalPath,
+    },
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+use serde_json::Value;
+use typenum::Unsigned;
+
+use std::{cmp::Ordering, sync::Arc};
+
+// This trait defines methods that a data source should implement
+// It enables retrieval of the membership path for a leaf node, which can be used to reconstruct the Merkle tree state.
+#[async_trait]
+pub trait MerklizedStateDataSource<Types>
+where
+    Types: NodeType,
+{
+    type Error: std::error::Error + std::fmt::Debug + Send + Sync + 'static;
+    async fn get_path<
```

Comment:
I think this is a bit messy and error prone. It makes it easy to accidentally mix different types of state in the same table, for example.

What I was imagining was that this _trait_ would be parameterized on the type of state being stored, and that in turn would tell you the types of the other generic parameter, the names of the tables and columns involved, etc. So e.g. we might have `trait MerklizedStateStorage<Types: NodeType, State: MerklizedState>`, and then add associated types to `MerklizedState` for `E` and `T` (I think it already has `I` and `A`).

This means this trait might be implemented twice in the sequencer, e.g. `impl MerklizedStateDataSource<SeqTypes, FeeMerkleTree> for SqlStorage` and `impl MerklizedStateDataSource<SeqTypes, BlocksMerkleTree> for SqlStorage`. This is fine IMO because we already have a customization point for updates (`UpdateStateData`).

Reply:
yeah this sounds way better